### PR TITLE
flesh out proposed solutions

### PR DESCRIPTION
**Pull request by chrisdickinson**

### CODE REVIEW

File: proposed-solutions.md [hunk 1]
```diff
@@ -1,35 +1,65 @@
...
+Determine the module type of the file by the presence or absence of a `"use
+module"` directive, defined and implemented *outside* of the JS specification.
 
-You write `js:;` at the beginning of a file to run as an ES Module.
+There are some issues with this approach:
 
-This eliminates the _major_ problem with the directives approach.
+* A circular problem arises: in order to accurately determine whether to parse
+  the file as an ES module or a Node.js module, one must first parse the file
+  looking for directive statements *(Consider that a directive may be proceeded
```
Comment (othiym23) [inline proposed-solutions.md]: s/proceeded/preceded/

File: proposed-solutions.md [hunk 2]
```diff
@@ -39,8 +69,48 @@ You write external metadata in a `package.json` to determine file type.
 }
 ```
 
-#### 6. extension (".mjs")
+More can be read about this proposal
+[here](https://github.com/dherman/defense-of-dot-js/blob/master/proposal.md).
+
+* While this removes all ambiguity at parse time, it does so by tightly
```
Comment (othiym23) [inline proposed-solutions.md]: not sure, but "tightly" could be construed as tendentious here

File: proposed-solutions.md [hunk 3]
```diff
@@ -39,8 +69,48 @@ You write external metadata in a `package.json` to determine file type.
...
+* Care must be taken in implementing `"modules.root"` to avoid allowing the
+  new package root to be set outside of the current package.
+* It does not solve for standalone JS files.
+
+#### 6. File Extension (".mjs")
+
+Use a new extension to indicate whether a file is an ES module (`.mjs`) or a
+Node.js module (`.js`).
+
+* This requires the registration of a new MIME type.
```
Comment (othiym23) [inline proposed-solutions.md]: Why? The browser loader spec determines the file type from how the module was requested, so browsers don't use MIME information. I know that @bradleymeck did this, but I don't know why it was required, so I doubt most other people do either.

File: proposed-solutions.md [hunk 4]
```diff
@@ -1,35 +1,65 @@
...
 
-#### 3. magic bytes ("js:;")
+Determine the module type of the file by the presence or absence of a `"use
+module"` directive, defined and implemented *outside* of the JS specification.
 
-You write `js:;` at the beginning of a file to run as an ES Module.
+There are some issues with this approach:
 
-This eliminates the _major_ problem with the directives approach.
+* A circular problem arises: in order to accurately determine whether to parse
```
Comment (othiym23) [inline proposed-solutions.md]: Not sure what's circular about this.

File: proposed-solutions.md (hunk 1 above)
Comment (othiym23) [inline proposed-solutions.md]: also, might be worth making explicit that this requires AST extraction and that regexps are insufficient

**Comment by othiym23** [comment]
:sheep::rocket::fire:

File: proposed-solutions.md (hunk 3 above)
Comment (bmeck) [inline proposed-solutions.md]: requires us to update the existing MIME in IANA (controlled by IESG, http://www.iana.org/assignments/media-types/application/javascript), posterity and people like chalker want it to be official if taken.

**Pull request merged**